### PR TITLE
synapse ws reconnect

### PR DESCRIPTION
**Pull request by VikrantKS**
# Issue Link

https://github.com/LambdaTest/test-at-scale/issues/41

# Description

This PR will add following functionality
- Add reconnect functionality with lambdatest server
- Handle auth failed and duplicate connection issue 

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Tested locally 
# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: cmd/synapse/bin.go [hunk 1]
```diff
@@ -138,6 +141,21 @@ func run(cmd *cobra.Command, args []string) {
 			case <-time.After(global.GracefulTimeout):
 				logger.Errorf("Graceful timeout exceeded. Brutally killing the application")
 			}
+		}
+
+	case <-connectionFailed:
```
Comment (palashmarantas) [inline cmd/synapse/bin.go:146]: how are re-trying case of connection failed ?

File: pkg/synapse/synapse.go [hunk 2]
```diff
@@ -5,30 +5,39 @@ import (
...
 const (
-	Repo    = "repo"
-	BuildID = "build-id"
-	JobID   = "job-id"
-	Mode    = "mode"
-	ID      = "id"
+	Repo                   = "repo"
+	BuildID                = "build-id"
+	JobID                  = "job-id"
+	Mode                   = "mode"
```
Comment (palashmarantas) [inline pkg/synapse/synapse.go:25]: what is mode here ?

File: pkg/synapse/synapse.go [hunk 3]
```diff
@@ -37,44 +46,102 @@ func New(
...
+		backoff.WithJitterFactor(0.05),                       // 5% jitter
+		backoff.WithMaxRetries(global.MaxConnectionAttempts), // If not specified, default number of retries is 10
+	)
+
+	normalCloser := make(chan struct{})
 
+	b, cancel := policy.Start(context.Background())
+	defer cancel()
 	s.logger.Debugf("starting socket connection")
 	s.logger.Errorf("starting socket connection at URL %s", global.SocketURL[viper.GetString("env")])
```
Comment (saurabh-prakash) [inline pkg/synapse/synapse.go:89]: stream-line this. Both debug and error logs?

File: cmd/synapse/bin.go (hunk 1 above)
Comment (VikrantKS) [reply cmd/synapse/bin.go:146]: This will be executed if max attempt to connection exhausted & failed

File: pkg/synapse/synapse.go (hunk 2 above)
Comment (VikrantKS) [reply pkg/synapse/synapse.go:25]: Mode defines the type of job , namely parsing , execution , coverage

File: pkg/synapse/synapse.go [hunk 4]
```diff
@@ -37,44 +46,102 @@ func New(
...
 }
 
 func (s *synapse) InitiateConnection(
 	ctx context.Context,
 	wg *sync.WaitGroup,
-) {
+	connectionFailed chan struct{}) {
 	defer wg.Done()
+	go s.openAndMaintainConnection(ctx, connectionFailed)
+	<-ctx.Done()
```
Comment (saurabh-prakash) [inline pkg/synapse/synapse.go:68]: not getting what you're trying to do here. Passing `ctx` to a go-routine and then immediately sending a Done signal. Won't it cause issues in go-routine where ctx is being used?

Comment (VikrantKS) [reply pkg/synapse/synapse.go:68]: `<-ctx.Done()` will receive channel message 
both  openAndMaintainConnection and InitiateConnection will listen  for done channel

**Review by palashmarantas — DISMISSED** [review]

File: pkg/global/synapseconstants.go [hunk 5]
```diff
@@ -6,20 +6,21 @@ import (
...
-	RepoSecretsFileName  = "reposecrets"
-	SynapseContainerURL  = "http://synapse:8000"
-	NetworkEnvName       = "NetworkName"
-	AutoRemoveEnv        = "AutoRemove"
-	SynapseHostEnv       = "synapsehost"
-	LocalEnv             = "local"
-	NetworkName          = "test-at-scale"
-	AutoRemove           = true
-	Local                = true
+	GracefulTimeout       = 100000 * time.Millisecond
```
Comment (rishabh-arya95) [inline pkg/global/synapseconstants.go:9]: nit: Can't we use seconds instead of ms

Comment (VikrantKS) [reply pkg/global/synapseconstants.go]: noted and changed

**Review by palashmarantas — APPROVED** [review]

**Review by rishabh-arya95 — APPROVED** [review]

**Pull request merged**